### DISCUSSION TITLE
Internal ESP32 DAC

### DISCUSSION THREAD

**Discussion by dkm1978** [schreibfaul1/ESP32-audioI2S#95] [General]
Hello.
I fork your's ESP32-audioI2S
I added internal DAC support
It works very well :)

https://www.youtube.com/watch?v=RMBcPz6vkwY

**Comment by schreibfaul1** [comment]
Hello kdm1978, good video! In many cases the 8-bit resolution of the internal DAC may be sufficient. Internally, the decoders work with a 16-bit resolution. The sound quality can be improved with an external DAC (for example PCM5102).

**Reply by dkm1978** [reply]
Thank you for your answer.
I know that the built-in converters are 8 bit but for me it is a sufficient resolution :)

**Reply by morgan-wild** [reply]
The sound you have seems great for a 8-bit DAC!
I also tried withour an external DAC and it was horrible!

**Comment by mettyw** [comment]
Hi,

I am trying to use this lib with the internal DAC of an M5Stack Grey. At least for me, the changes in `m_i2s_config` aren't sufficient, I hear sound but it is extremely distorted. I could fix it by altering the function `Audio::Gain()` as well and add something like 
`v[LEFTCHANNEL] = v[LEFTCHANNEL]  + 0x8000;`
`v[RIGHTCHANNEL] = v[RIGHTCHANNEL]  + 0x8000;`
similar to what has been done here: https://github.com/earlephilhower/ESP8266Audio/blob/6ecb6afe3059375c19377cd94a2c99f83c64e3c8/src/AudioOutputI2S.cpp#L238 do you think it is possible to extend your code accordingly? 

My understanding of ESP32, C and bit-pushing is a bit rusty, so sorry for not providing a clearer fix directly.

**Reply by mettyw** [reply]
Ah I forgot to mention that I also had to change the call to setPinout(). Actually, now I realize that all my changes are also there in @dkm1978 's original fork:
* https://github.com/dkm1978/ESP32-audioI2S/blob/5f3e30d0b3c61c48b81506545d665bbe4eb7936b/src/Audio.cpp#L143
* https://github.com/dkm1978/ESP32-audioI2S/blob/5f3e30d0b3c61c48b81506545d665bbe4eb7936b/src/Audio.cpp#L2149 
* https://github.com/dkm1978/ESP32-audioI2S/blob/5f3e30d0b3c61c48b81506545d665bbe4eb7936b/src/Audio.cpp#L2261

So, I rephrase my question: would it be possible to include the above changes in a proper way so that internal DACs also work smoothly?

**Comment by dkm1978** [comment]
Add #define INTDAC in audio. Cpp 
  Dnia 01 lutego 2021 22:02 mettyw   napisał(a):  
 Hi, 
 I am trying to use this lib with the internal DAC of an M5Stack Grey. At least for me, the changes in  m_i2s_config  aren&#39;t sufficient, I hear sound but it is extremely distorted. I could fix it by altering the function  Audio::Gain()  as well and add something like  
 v[LEFTCHANNEL] = v[LEFTCHANNEL]  + 0x8000;  
 v[RIGHTCHANNEL] = v[RIGHTCHANNEL]  + 0x8000;  
similar to what has been done here:  github.com github.com  do you think it is possible to extend your code accordingly? 
 My understanding of C and bit-pushing is a bit rusty, so sorry for not providing a clearer fix directly.

**Comment by dkm1978** [comment]
See at 2261 line in my audio.cpp  
  Dnia 01 lutego 2021 22:02 mettyw   napisał(a):  
 Hi, 
 I am trying to use this lib with the internal DAC of an M5Stack Grey. At least for me, the changes in  m_i2s_config  aren&#39;t sufficient, I hear sound but it is extremely distorted. I could fix it by altering the function  Audio::Gain()  as well and add something like  
 v[LEFTCHANNEL] = v[LEFTCHANNEL]  + 0x8000;  
 v[RIGHTCHANNEL] = v[RIGHTCHANNEL]  + 0x8000;  
similar to what has been done here:  github.com github.com  do you think it is possible to extend your code accordingly? 
 My understanding of C and bit-pushing is a bit rusty, so sorry for not providing a clearer fix directly.

**Reply by mettyw** [reply]
Thanks for the hints! I also now realized that your fork already contains all required.changes.

However, I am not using this Lib here directly,. Instead, I am playing around with https://github.com/biologist79/ESPuino and therefore it would be really cool if @schreibfaul1 could integrate this into his version.

Furthermore, now with my increased understanding I think the instructions here in the wiki are incomplete, since they only describe the changes to the  initialization of i2s

**Reply by schreibfaul1** [reply]
setInternalDAC(true); //enables internal DAC

**Reply by ranger2012** [reply]
I 've compared both codes from  dkm1978 and  from Wolly, it seems that  dkm1978  source is much better, maybe   somewhere else in the code there is difference...

**Reply by ranger2012** [reply]
setting the volume to maximum  makes the sound clearer..

**Reply by mettyw** [reply]
> setInternalDAC(true); //enables internal DAC

Hey thanks for adding this function and the flexible way how it is implemented is even cooler :)

However, there is still heavy noise using this lib compared to other libs. Actually, this can be avoided: I saw that you decided to add the increment before the call of the `Gain()` function. I think I understand the reasoning for that, considering the volume, filter & balance calculations that are done within `Gain()` . But according to https://github.com/espressif/arduino-esp32/issues/2647#issuecomment-482713183 the purpose is to convert from signed to unsigned, since the internal DAC expects unsigned. So, I think the increment of 0x8000 should be at the very end before `Gain()` returns the result, regardless of the calculations that happened before.

If I do that change in my code, the output of the DAC is actually really clear (... well, for an internal DAC at least).

**Reply by schreibfaul1** [reply]
Hi mettyw,
that's it! Whether the output signal is distorted or not depends on the volume. The offset 0x8000 is now added after Gain (). Many Thanks!

**Reply by ranger2012** [reply]
sound is good now.. Thank you both for such a wonderful work..
